pkg/operator/ceph/cluster/cluster.go

```diff
@@ -138,16 +137,14 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
 }
 
 func (c *cluster) createInstance(rookImage string, cephVersion cephver.CephVersion) error {
-	if c.checkSetOrchestrationRunning() == true {
-		logger.Debugf("As createInstance is currently running added this request as pending.")
-		c.setOrchestrationPending()
-		return nil
-	}
-
-	defer c.unsetOrchestrationRunning()
-	initRun := true
-	for c.checkUnsetOrchestrationPending() == true || initRun == true {
-		initRun = false
+	// isRunningThread indicates if the calling thread is the one which executes the for-loop
+	isRunningThread := false
+	c.setOrchestrationNeeded()
+
```

Comment:
To ensure the orchestration status is cleared, should we call this here?
```
defer c.unsetOrchestrationStatus()
```

Reply:
This would re-introduce the race condition I pointed out in the previous iteration.

Reply:
Ha, good point, let's avoid that...

To avoid all the lines where `unsetOrchestrationStatus()` is called, how about if we split the contents of the `for` loop out to another method (approx lines 148-216)? If that method returns an error, then `unsetOrchestrationStatus()` is called. This way, the code controlling the locking is all in a single method and the orchestration logic is all in other methods.